
Always declare path specification in case of inconsistent declaration
elsewhere. (bsc#1209165)

---
 util/grub-mkconfig.in       |    3 ++-
 util/grub-mkconfig_lib.in   |    4 ++++
 util/grub.d/00_header.in    |   23 ++++++++++++++++++++++-
 util/grub.d/10_linux.in     |   11 ++++++++++-
 util/grub.d/20_linux_xen.in |    4 ++++
 5 files changed, 42 insertions(+), 3 deletions(-)

Index: grub-2.16/util/grub-mkconfig_lib.in
===================================================================
--- grub-2.16.orig/util/grub-mkconfig_lib.in
+++ grub-2.16/util/grub-mkconfig_lib.in
@@ -49,7 +49,11 @@ grub_warn ()
 
 make_system_path_relative_to_its_root ()
 {
+  if [ "x${SUSE_BTRFS_SNAPSHOT_BOOTING}" = "xtrue" ] ; then
+  "${grub_mkrelpath}" -r "$1"
+  else
   "${grub_mkrelpath}" "$1"
+  fi
 }
 
 is_path_readable_by_grub ()
Index: grub-2.16/util/grub.d/00_header.in
===================================================================
--- grub-2.16.orig/util/grub.d/00_header.in
+++ grub-2.16/util/grub.d/00_header.in
@@ -27,6 +27,21 @@ export TEXTDOMAINDIR="@localedir@"
 
 . "$pkgdatadir/grub-mkconfig_lib"
 
+if [ "x${SUSE_BTRFS_SNAPSHOT_BOOTING}" = "xtrue" ] &&
+   [ "x${GRUB_FS}" = "xbtrfs" ] ; then
+    cat <<EOF
+set btrfs_relative_path="y"
+export btrfs_relative_path
+EOF
+else
+# Always declare path specification in case of inconsistent declaration
+# elsewhere. (bsc#1209165)
+    cat <<EOF
+set btrfs_relative_path="n"
+export btrfs_relative_path
+EOF
+fi
+
 # Do this as early as possible, since other commands might depend on it.
 # (e.g. the `loadfont' command might need lvm or raid modules)
 for i in ${GRUB_PRELOAD_MODULES} ; do
@@ -90,7 +105,9 @@ or please file a bug report." >&2
 fi
 
 cat << EOF
-if [ -s \$prefix/grubenv ]; then
+if [ -f \${config_directory}/grubenv ]; then
+  load_env -f \${config_directory}/grubenv
+elif [ -s \$prefix/grubenv ]; then
   load_env
 fi
 
@@ -491,3 +508,15 @@ fi
 if [ "x${GRUB_BADRAM}" != "x" ] ; then
   echo "badram ${GRUB_BADRAM}"
 fi
+
+if [ "x${SUSE_BTRFS_SNAPSHOT_BOOTING}" = "xtrue" ] &&
+   [ "x${GRUB_FS}" = "xbtrfs" ] ; then
+    # Note: No $snapshot_num on *read-only* rollback!  (bsc#901487)
+    cat <<EOF
+if [ -n "\$extra_cmdline" ]; then
+  submenu "Bootable snapshot #\$snapshot_num" {
+    menuentry "If OK, run 'snapper rollback' and reboot." { true; }
+  }
+fi
+EOF
+fi
Index: grub-2.16/util/grub-mkconfig.in
===================================================================
--- grub-2.16.orig/util/grub-mkconfig.in
+++ grub-2.16/util/grub-mkconfig.in
@@ -322,7 +322,8 @@ export GRUB_DEFAULT \
   GRUB_ADVANCED_ACCESS_LEVEL \
   GRUB_RECOVERY_ACCESS_LEVEL \
   GRUB_BOOTNEXT_OMIT_INACTIVE \
-  GRUB_DISABLE_MODULE_DEDUPLICATION
+  GRUB_DISABLE_MODULE_DEDUPLICATION \
+  SUSE_BTRFS_SNAPSHOT_BOOTING
 
 if test "x${grub_cfg}" != "x"; then
   rm -f "${grub_cfg}.new"
Index: grub-2.16/util/grub.d/10_linux.in
===================================================================
--- grub-2.16.orig/util/grub.d/10_linux.in
+++ grub-2.16/util/grub.d/10_linux.in
@@ -71,10 +71,14 @@ fi
 
 case x"$GRUB_FS" in
     xbtrfs)
+	if [ "x${SUSE_BTRFS_SNAPSHOT_BOOTING}" = "xtrue" ]; then
+	GRUB_CMDLINE_LINUX="${GRUB_CMDLINE_LINUX} \${extra_cmdline}"
+	else
 	rootsubvol="`make_system_path_relative_to_its_root /`"
 	rootsubvol="${rootsubvol#/}"
 	if [ "x${rootsubvol}" != x ]; then
 	    GRUB_CMDLINE_LINUX="rootflags=subvol=${rootsubvol} ${GRUB_CMDLINE_LINUX}"
+	fi
 	fi;;
     xzfs)
 	rpool=`${grub_probe} --device ${GRUB_DEVICE} --target=fs_label 2>/dev/null || zdb -l ${GRUB_DEVICE} | awk -F \' '/ name/ { print $2 }'`
@@ -306,7 +310,12 @@ for linux in ${reverse_sorted_list}; do
   if [ $PLATFORM != "emu" ]; then
     hotkey=0
   else
-    rel_dirname=$dirname
+    if [ "x${SUSE_BTRFS_SNAPSHOT_BOOTING}" = "xtrue" ] &&
+       [ "x${GRUB_FS}" = "xbtrfs" ] ; then
+       rel_dirname="\${btrfs_subvol}$dirname"
+    else
+       rel_dirname="$dirname"
+    fi
   fi
   version=`echo $basename | sed -e "s,^[^0-9]*-,,g"`
   alt_version=`echo $version | sed -e "s,\.old$,,g"`
Index: grub-2.16/util/grub.d/20_linux_xen.in
===================================================================
--- grub-2.16.orig/util/grub.d/20_linux_xen.in
+++ grub-2.16/util/grub.d/20_linux_xen.in
@@ -81,10 +81,14 @@ fi
 
 case x"$GRUB_FS" in
     xbtrfs)
+	if [ "x${SUSE_BTRFS_SNAPSHOT_BOOTING}" = "xtrue" ]; then
+	GRUB_CMDLINE_LINUX="${GRUB_CMDLINE_LINUX} \${extra_cmdline}"
+	else
 	rootsubvol="`make_system_path_relative_to_its_root /`"
 	rootsubvol="${rootsubvol#/}"
 	if [ "x${rootsubvol}" != x ]; then
 	    GRUB_CMDLINE_LINUX="rootflags=subvol=${rootsubvol} ${GRUB_CMDLINE_LINUX}"
+	fi
 	fi;;
     xzfs)
 	rpool=`${grub_probe} --device ${GRUB_DEVICE} --target=fs_label 2>/dev/null || zdb -l ${GRUB_DEVICE} | awk -F \' '/ name/ { print $2 }'`
